PyInstaller/hooks/hook-pkg_resources.py

```diff
@@ -9,26 +9,33 @@
 # SPDX-License-Identifier: (GPL-2.0-or-later WITH Bootloader-exception)
 #-----------------------------------------------------------------------------
 
-from PyInstaller.utils.hooks import collect_submodules, is_module_satisfies
+from PyInstaller.utils.hooks import collect_submodules, is_module_satisfies, can_import_module, is_module_or_submodule
 
 # pkg_resources keeps vendored modules in its _vendor subpackage, and does sys.meta_path based import magic to expose
 # them as pkg_resources.extern.*
-hiddenimports = collect_submodules('pkg_resources._vendor')
+
+# The `railroad` package is an optional requirement for `pyparsing`. `pyparsing.diagrams` depends on `railroad`, so
+# filter it out when `railroad` is not available.
+if can_import_module('railroad'):
+    hiddenimports = collect_submodules('pkg_resources._vendor')
+else:
+    hiddenimports = collect_submodules(
+        'pkg_resources._vendor',
+        filter=lambda name: is_module_or_submodule(name, 'pkg_resources._vendor.pyparsing.diagrams')
```

Comment:
I think the filter lambda is missing a `not`? Since `True` means that we want to include the sub-package.

Also, I don't think we need `is_module_or_submodule` anymore. A `lambda name: name != 'pkg_resources._vendor.pyparsing.diagrams'` should suffice, as we do not recurse into excluded sub-packages anymore. (In earlier implementations, we recursed into all sub-packages and then filtered out the results, which is why `is_module_or_submodule` was necessary - to filter out both the excluded sub-package and its contents...)

Reply:
Thanks @rokm for the feedback, that is great! I have updated the PR with that change.